### PR TITLE
Add CloudConfig operations

### LINKED ISSUE #588: TODO: Add CloudConfig to support cluster creation in cloud environments
In order to move toward supporting cloud environments, we need to add CloudConfig to Zookeeper. This config can be used for participant auto registration.

### PR DESCRIPTION
**Pull request by alirezazamani**
### Issues
- [x] My PR addresses the following Helix issues and references them in the PR title:
Fixes #588 .

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
In order to move toward supporting cloud environments, we need to add CloudConfig to Zookeeper. The CloudConfig information will be used by the participants for autojoining. 
In this commit:
The code regarding CloudConfig is added. A new test has been added to check the correctness of the code.

### Tests

- [x] The following is the result of the "mvn test" command on the appropriate module:
A new test has been added. (TestCloudConfig)

Test Result 1: mvn test
[ERROR] Tests run: 883, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 3,234.951 s <<< FAILURE! - in TestSuite
[ERROR] testTaskPerformanceMetrics(org.apache.helix.monitoring.mbeans.TestTaskPerformanceMetrics)  Time elapsed: 2.22 s  <<< FAILURE!
java.lang.AssertionError: expected:<true> but was:<false>
	at org.apache.helix.monitoring.mbeans.TestTaskPerformanceMetrics.testTaskPerformanceMetrics(TestTaskPerformanceMetrics.java:118)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestTaskPerformanceMetrics.testTaskPerformanceMetrics:118 expected:<true> but was:<false>
[INFO] 
[ERROR] Tests run: 883, Failures: 1, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  53:59 min
[INFO] Finished at: 2019-11-12T15:32:39-08:00
[INFO] ------------------------------------------------------------------------

Test Result 2:  mvn test -Dtest="TestTaskPerformanceMetrics"
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 16.397 s - in org.apache.helix.monitoring.mbeans.TestTaskPerformanceMetrics
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  21.497 s
[INFO] Finished at: 2019-11-12T15:46:06-08:00
[INFO] ------------------------------------------------------------------------

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] My diff has been formatted using helix-style.xml

### CODE REVIEW

**Comment by alirezazamani** [comment]
@dasahcc Could you please review this PR based on our discussions.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+/**
+ * Cloud configurations
+ */
+public class CloudConfig extends HelixProperty {
+  /**
+   * Configurable characteristics of a cloud.
+   * NOTE: Do NOT use this field name directly, use its corresponding getter/setter in the
+   * CloudConfig.
+   */
+  public enum CloudConfigProperty {
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: A bit confused about this. If you say "Do NOT use this field name directly", is this enum supposed to be private?

File: helix-core/src/test/java/org/apache/helix/model/cloud/TestCloudConfig.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    // Write the CloudConfig to Zookeeper
+    ZKHelixDataAccessor accessor =
+        new ZKHelixDataAccessor(clusterName, new ZkBaseDataAccessor(_gZkClient));
+    Builder keyBuilder = accessor.keyBuilder();
+    accessor.setProperty(keyBuilder.cloudConfig(), cloudConfig);
+
+    // Read CloudConfig from Zookeeper and check the content
+    ConfigAccessor _configAccessor = new ConfigAccessor(_gZkClient);
+    CloudConfig cloudConfigFromZk = _configAccessor.getCloudConfig(clusterName);
```
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/model/cloud/TestCloudConfig.java]: Can you add test to cover exception and null in `getCloudConfig()`? We try to cover as much as possible in tests.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    CLOUD_ENABLED, // determine whether the cluster is inside cloud environment.
+    CLOUD_ID, // the cloud Id that belongs to this cluster.
+    TOPOLOGY_URL, // the url from where to retrieve the topology information.
+    TOPOLOGY_PARSER, // the name of the function that parses the topology.
```
Comment (zhangmeng916) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Let's make the two fields starting with "TOPOLOGY" be more generic. The fetcher and parser actually handle more information than just topology.

File: helix-core/src/test/java/org/apache/helix/model/cloud/TestCloudConfig.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
+package org.apache.helix.model.cloud;
```
Comment (zhangmeng916) [inline helix-core/src/test/java/org/apache/helix/model/cloud/TestCloudConfig.java]: You may want to check TestHelixConfigAccessor and TestConfigAccessor. You can merge your test into the existing tests, as cloud config is comparable to cluster config, participant config etc, and the test environment is already set up there.

Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/model/cloud/TestCloudConfig.java]: I would say let's keep them separated because this config is newly introduced and it is easier to maintain it later. Perhaps we will add more test to this Tests later to this file.

File: helix-core/src/main/java/org/apache/helix/ConfigAccessor.java [hunk 5]
```diff
@@ -562,6 +563,30 @@ public ClusterConfig getClusterConfig(String clusterName) {
...
+  public CloudConfig getCloudConfig(String clusterName) {
+    if (!ZKUtil.isClusterSetup(clusterName, zkClient)) {
+      throw new HelixException("fail to get config. cluster: " + clusterName + " is NOT setup.");
+    }
+    HelixConfigScope scope =
+        new HelixConfigScopeBuilder(ConfigScopeProperty.CLOUD).forCluster(clusterName).build();
+    ZNRecord record = getConfigZnRecord(scope);
+
+    if (record == null) {
+      LOG.warn("No cloud config found at " + scope.getZkPath());
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: Done.

File: helix-core/src/main/java/org/apache/helix/ConfigAccessor.java [hunk 6]
```diff
@@ -562,6 +563,30 @@ public ClusterConfig getClusterConfig(String clusterName) {
     return new ClusterConfig(record);
   }
 
+
+  /**
+   * Get CloudConfig of the given cluster.
+   *
+   * @param clusterName
+   *
+   * @return
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: Done.

File: helix-core/src/test/java/org/apache/helix/model/cloud/TestCloudConfig.java (hunk 2 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/model/cloud/TestCloudConfig.java]: Done.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 3 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Done.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 1 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: I changed it to private.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+public class CloudConfig extends HelixProperty {
+  /**
+   * Configurable characteristics of a cloud.
+   * NOTE: Do NOT use this field name directly, use its corresponding getter/setter in the
+   * CloudConfig.
+   */
+  private enum CloudConfigProperty {
+    CLOUD_ENABLED, // determine whether the cluster is inside cloud environment.
+    CLOUD_ID, // the cloud Id that belongs to this cluster.
+    CLOUD_INFO_FETCHER, // the url from where to retrieve the cloud information.
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Fetcher sounds a little bit confusing. Let's use CLOUD_INFO_URL to make it more clear.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    CLOUD_INFO_PARSER, // the name of the function that parses the topology.
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Same there. It is better for CLOUD_INFO_PARSE_CLASS_NAME. It will align with IdealState's REBALANCER_CLASS_NAME

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  /**
+   * Instantiate the CloudConfig for the cloud
+   * @param cluster
+   */
+  public CloudConfig(String cluster) {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: We can think about the builder here, since we need to verify the completion of cloud config. For example, if you enable cloud config but there is no URL or parser class. Shall we fail it?

For details, you can refer WorkflowConfig.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 7 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Fixed.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 8 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Fixed.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 9 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Updated.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+  /**
+   * Configurable characteristics of a cloud.
+   * NOTE: Do NOT use this field name directly, use its corresponding getter/setter in the
+   * CloudConfig.
+   */
+  public enum CloudConfigProperty {
+    CLUSTER_NAME,   // name of the cluster associated with this
+    CLOUD_ENABLED, // determine whether the cluster is inside cloud environment.
+    CLOUD_ID, // the cloud Id that belongs to this cluster.
+    CLOUD_INFO_URL, // the url from where to retrieve the cloud information.
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: URLS?

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+ * Cloud configurations
+ */
+public class CloudConfig extends HelixProperty {
+  /**
+   * Configurable characteristics of a cloud.
+   * NOTE: Do NOT use this field name directly, use its corresponding getter/setter in the
+   * CloudConfig.
+   */
+  public enum CloudConfigProperty {
+    CLUSTER_NAME,   // name of the cluster associated with this
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: This is not necessary. Your ZNode id is your cluster name.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+  /**
+   * Instantiate the CloudConfig for the cloud
+   * @param cluster
+   */
+  public CloudConfig(String cluster) {
+    super(cluster);
+  }
+
+
+  protected CloudConfig(String cluster, boolean enabled, String cloudID, List<String> cloudInfoURL,
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: cloudInfoURLs?

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+   */
+  public void setCloudEnabled(boolean enabled) {
+    _record.setBooleanField(CloudConfigProperty.CLOUD_ENABLED.name(), enabled);
+  }
+
+  /**
+   * Whether CLOUD_ENABLED field is enabled or not.
+   * @return
+   */
+  public boolean getCloudEnabled() {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: isCloudEnabled?

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+   */
+  public String getCloudID() {
+    return _record.getSimpleField(CloudConfigProperty.CLOUD_ID.name());
+  }
+
+  /**
+   * Set the CLOUD_INFO_URL field.
+   * @param cloudInfoURL
+   */
+  public void setCloudInfoURL(List<String> cloudInfoURL) {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Same here. URLs

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+
+    private void validate() {
+      if (_clusterName == null) {
+        throw new HelixException("The cluster name/ID cannot be null.");
+      }
+
+      if (_cloudEnabled) {
+        if (_cloudID == null || _cloudInfoURLList == null || _cloudInfoParserClassName == null || _cloudInfoURLList.isEmpty()) {
+          throw new HelixException(
+              "This Cloud Configuration is Invalid. The configuration must have all of the fields if enabled.");
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: This could be confusing here. How the user know the all fields?

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 10 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Fixed. Thanks.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 11 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Addressed.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 12 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Fixed.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 13 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Done. Thanks.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 14 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Done.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 15 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Addressed. Thanks.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 16]
```diff
@@ -36,10 +36,9 @@
    * CloudConfig.
    */
   public enum CloudConfigProperty {
-    CLUSTER_NAME,   // name of the cluster associated with this
     CLOUD_ENABLED, // determine whether the cluster is inside cloud environment.
     CLOUD_ID, // the cloud Id that belongs to this cluster.
-    CLOUD_INFO_URL, // the url from where to retrieve the cloud information.
+    CLOUD_INFO_URLS, // the url from where to retrieve the cloud information.
```
Comment (zhangmeng916) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Change the "url" to "urls" in comment

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 17]
```diff
@@ -167,21 +156,32 @@ public CloudConfig build() {
...
+    /**
+     * Constructor with Cluster Name as input
+     * @param clusterName
+     */
     public Builder(String clusterName) {
       _clusterName = clusterName;
     }
 
+    /**
+     * onstructor with CloudConfig as input
```
Comment (zhangmeng916) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: typo

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 18]
```diff
@@ -238,14 +237,19 @@ public String getCloudInfoParserClassName () {
...
+        if (_cloudID == null ) {
+          throw new HelixException(
+              "This Cloud Configuration is Invalid. The CloudID is missing from the config.");
+        } else if ( _cloudInfoURLList == null) {
+          throw new HelixException(
+              "This Cloud Configuration is Invalid. The CloudInfoURLList is missing from the config.");
+        } else if (_cloudInfoParserClassName == null) {
+          throw new HelixException(
+              "This Cloud Configuration is Invalid. The CloudInfoParserClassName is missing from the config.");
+        } else if (_cloudInfoURLList.isEmpty()) {
```
Comment (zhangmeng916) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Combine this check with the previous cloudInfoURLList check to make it less confusing.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 16 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Good catch. Done.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 17 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Fixed.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 18 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Good suggestion. Fixed.

File: helix-core/src/main/java/org/apache/helix/PropertyKey.java [hunk 19]
```diff
@@ -214,6 +215,16 @@ public PropertyKey clusterConfig() {
...
     }
 
+
+    /**
+     * Get a property key associated with this Cloud configuration
+     * @return {@link PropertyKey}
+     */
+    public PropertyKey cloudConfig() {
+      return new PropertyKey(CONFIGS, ConfigScopeProperty.CLOUD, CloudConfig.class,
+          _clusterName, ConfigScopeProperty.CLOUD.toString(), _clusterName);
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/PropertyKey.java]: Dont use toString(). Our convention will be use the enum's name() function. Also please change the previous one.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+      _cloudID = v;
+      return this;
+    }
+
+    public Builder setCloudInfoURL (List<String> v) {
+      _cloudInfoURLList = v;
+      return this;
+    }
+
+    public Builder setCloudInfoURLField (String v) {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: If you prefer to provide function to put one URL at a time, let's call it addCloudInfoURL.

File: helix-core/src/main/java/org/apache/helix/PropertyKey.java (hunk 19 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/PropertyKey.java]: Fixed.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 20 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Done.

**Comment by alirezazamani** [comment]
This PR is ready to be merged, approved by @dasahcc.

Title: Add CloudConfig Related Code

Body:    
In order to move toward supporting cloud environments and auto-registration,
we need to have a CloudConfig Znode under the Config. 
In this commit, the code regarding the CloudConfig is added.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+  public enum CloudConfigProperty {
+    CLOUD_ENABLED, // determine whether the cluster is inside cloud environment.
+    CLOUD_ID, // the cloud Id that belongs to this cluster.
+    CLOUD_INFO_URLS, // the URLs from where to retrieve the cloud information.
+    CLOUD_INFO_FETCHER_CLASS_NAME, // the name of the function that fetches the topology.
+    CLOUD_INFO_PARSER_CLASS_NAME // the name of the function that parses the topology.
+  }
+
+  /* Default values */
+  public static final boolean DEFAULT_CLOUD_ENABLED = false;
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Why is this public?

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 22]
```diff
@@ -0,0 +1,287 @@
...
+    public String getCloudInfoFetcherClassName() {
+      return _cloudInfoFetcherClassName;
+    }
+
+    public String getCloudInfoParserClassName() {
+      return _cloudInfoParserClassName;
+    }
+
+    private void validate() {
+      if (_cloudEnabled) {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: So here, users could create invalid CloudConfig objects/ZNodes at will as long as they set `_cloudEnabled` to false. I don't know if that makes too much sense. Is that intended?

File: helix-core/src/test/java/org/apache/helix/model/cloud/TestCloudConfig.java [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+
+
+
+  @Test(expectedExceptions = HelixException.class)
+  public void testUnverifiedCloudConfigBuilderEmptyURL() {
+    String className = getShortClassName();
+    String clusterName = "CLUSTER_" + className;
+    CloudConfig.Builder builder = new CloudConfig.Builder(clusterName);
+    builder.setCloudEnabled(true);
+    //
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/model/cloud/TestCloudConfig.java]: ?

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 24]
```diff
@@ -0,0 +1,287 @@
...
+  /**
+   * Instantiate with a pre-populated record
+   * @param record a ZNRecord corresponding to a cloud configuration
+   */
+  public CloudConfig(ZNRecord record) {
+    super(record);
+  }
+
+  /**
+   * Instantiate the config using each field individually.
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Add here that users should use the Builder instead.

File: helix-core/src/main/java/org/apache/helix/ConfigAccessor.java [hunk 25]
```diff
@@ -562,11 +563,33 @@ public ClusterConfig getClusterConfig(String clusterName) {
...
+   * Get CloudConfig of the given cluster.
    *
-   * @param clusterName The cluster
+   * @param clusterName
    *
+   * @return The instance of {@link CloudConfig}
+   */
+  public CloudConfig getCloudConfig(String clusterName) {
+    if (!ZKUtil.isClusterSetup(clusterName, zkClient)) {
+      throw new HelixException("Failed to get config. cluster: " + clusterName + " is NOT setup.");
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: Nit: String.format where applicable.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 21 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Fixed.

File: helix-core/src/test/java/org/apache/helix/model/cloud/TestCloudConfig.java (hunk 23 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/model/cloud/TestCloudConfig.java]: Removed.

File: helix-core/src/main/java/org/apache/helix/ConfigAccessor.java (hunk 25 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/ConfigAccessor.java]: Fixed. Thanks.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 24 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Added.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 22 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Yes. It is intended. This feature is enabled once CLOUD_ENABLED is true. If it is false, the feature is disabled.

Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: I am not referring to whether a feature is enabled or not; rather, I'm pointing to whether CloudConfig object created is valid or not. These two are separate things. Shouldn't the validation logic validate regardless of whether the feature is enabled or not?

Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: You want to have a check for all of the fields? If other fields are null or not? As I see in other builders such as workflow config we do not check all the fields. Do we?
@dasahcc what do you think?

Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: User should have the ability to create an empty CLOUDCONFIG before updating something. So if they are doing so, then it means they give up advantage of Helix validation.

File: helix-core/src/main/java/org/apache/helix/cloud/CloudProvider.java [hunk 26]
```diff
@@ -0,0 +1,25 @@
+package org.apache.helix.cloud;
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/cloud/CloudProvider.java]: Let's create a constant folder and move this to the /helix/could/constants, since we will have other constants there.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 27]
```diff
@@ -0,0 +1,289 @@
...
+  /**
+   * Configurable characteristics of a cloud.
+   * NOTE: Do NOT use this field name directly, use its corresponding getter/setter in the
+   * CloudConfig.
+   */
+  public enum CloudConfigProperty {
+    CLOUD_ENABLED, // determine whether the cluster is inside cloud environment.
+    CLOUD_PROVIDER, // the environment the cluster is in, e.g. Azure, AWS, or Customized
+    CLOUD_ID, // the cloud Id that belongs to this cluster.
+    CLOUD_INFO_SOURCE, // the source for retrieving the cloud information.
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Add some comments here. If user use Helix supported default provider, these two entries will not be shown in CLOUDCONFIG.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 28]
```diff
@@ -0,0 +1,289 @@
...
+   */
+  public CloudConfig(String cluster, boolean enabled, CloudProvider cloudProvider, String cloudID,
+      List<String> cloudInfoSource, String cloudProcessorName) {
+    super(cluster);
+    _record.setBooleanField(CloudConfigProperty.CLOUD_ENABLED.name(), enabled);
+    _record.setSimpleField(CloudConfigProperty.CLOUD_PROVIDER.name(), cloudProvider.name());
+    _record.setSimpleField(CloudConfigProperty.CLOUD_ID.name(), cloudID);
+    _record.setSimpleField(CloudConfigProperty.CLOUD_INFO_PROCESSOR_NAME.name(),
+        cloudProcessorName);
+    _record.setListField(CloudConfigProperty.CLOUD_INFO_SOURCE.name(), cloudInfoSource);
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Do a check here. only add these two entries when cloudProvider is CUSTOM

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 29]
```diff
@@ -0,0 +1,289 @@
...
+        }
+        if (_cloudProvider == null) {
+          throw new HelixException(
+              "This Cloud Configuration is Invalid. The Cloud Provider is missing from the config.");
+        } else if (_cloudProvider == CloudProvider.CUSTOM) {
+          if (_cloudInfoProcessorName == null) {
+            throw new HelixException(
+                "This Cloud Configuration is Invalid. Custom provider has been chosen without defining CloudInfoProcessorName.");
+          }
+          if (_cloudInfoSources == null || _cloudInfoSources.size() == 0) {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Combing these two. We just need let user know either processor or source is invalid.

File: helix-core/src/main/java/org/apache/helix/cloud/CloudProvider.java (hunk 26 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/cloud/CloudProvider.java]: Good suggestion. Done.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 28 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Done. Thanks.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 29 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Done.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 27 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Done.

**Comment by alirezazamani** [comment]
This PR is ready to be merged, approved by @dasahcc.

Title: Add CloudConfig Related Code

Body:
In order to move toward supporting cloud environments and auto-registration,
we need to have a CloudConfig Znode under the Config.
In this commit, the code regarding the CloudConfig is added.

File: helix-core/src/main/java/org/apache/helix/cloud/constants/CloudProvider.java [hunk 30]
```diff
@@ -0,0 +1,25 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+public enum CloudProvider {
+  AZURE,
+  CUSTOM
```
Comment (zhangmeng916) [inline helix-core/src/main/java/org/apache/helix/cloud/constants/CloudProvider.java]: CUSTOMIZED? "CUSTOM" seems not what you actually mean here.

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java [hunk 31]
```diff
@@ -0,0 +1,290 @@
...
+   */
+  public enum CloudConfigProperty {
+    CLOUD_ENABLED, // determine whether the cluster is inside cloud environment.
+    CLOUD_PROVIDER, // the environment the cluster is in, e.g. Azure, AWS, or Customized
+    CLOUD_ID, // the cloud Id that belongs to this cluster.
+
+    // If user uses Helix supported default provider, the below entries will not be shown in
+    // CloudConfig.
+    CLOUD_INFO_SOURCE, // the source for retrieving the cloud information.
+    CLOUD_INFO_PROCESSOR_NAME // the name of the function that process the fetching and parsing of
```
Comment (zhangmeng916) [inline helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: process -> processes

File: helix-core/src/main/java/org/apache/helix/cloud/constants/CloudProvider.java (hunk 30 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/cloud/constants/CloudProvider.java]: Fixed,

File: helix-core/src/main/java/org/apache/helix/model/CloudConfig.java (hunk 31 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/model/CloudConfig.java]: Done.
